File: web/src/types/song.ts
type AudioMetadata = {
  sample_rate: number;
  channels: number;
  duration: number;
};

type VideoMetadata = {
  duration: number;
};

type ImageMetadata = {
  width: number;
  height: number;
};

export type Genre = {
  name: string;
};

export type Tag = {
  name: string;
  type: string;
  description?: string;
};

export type Singer = {
  name: string;
  aliases: string[];
};

export type SongSinger = {
  singer: Singer;
  role: "main" | "featured" | string;
};

export type MediaSource = {
  storage_type: "local" | "s3" | "youtube" | "soundcloud";
  url: string;
  file_type: "audio" | "video" | "art";
  format_type: "mp3" | "wav" | "flac" | "mp4" | "webm" | "png" | "jpg" | "jpeg";
  checksum?: string;
  audio_metadata?: AudioMetadata;
  video_metadata?: VideoMetadata;
  image_metadata?: ImageMetadata;
};

export type SongAlias = {
  name: string;
  language: string;
};

export type Song = {
  hash_id: string;
  title: string;
  description: string;
  type: "cover" | "original" | "remix";
  released_at: string | null;
  genres: Genre[];
  tags: Tag[];
  singers: SongSinger[];
  media_sources: MediaSource[];
  aliases: SongAlias[];
  version: number;
  song_group_hash_id: string;
};

export type SongMedia = {
  file_type: string;
  url: string;
};

export type SongThumbnail = {
  song_group_hash_id: string;
  title: string;
  cover_url: string;
};

export type SongThumbnailWindow = {
  total_songs: number;
  songs: SongThumbnail[];
};

File: server/models/song.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type SongType string

const (
	SongTypeOriginal    SongType = "original"
	SongTypeCover       SongType = "cover"
	SongTypeRemix       SongType = "remix"
	SongTypeCompilation SongType = "compilation"
)

type Song struct {
	gorm.Model
	HashID          string   `gorm:"unique;not null" json:"hash_id"`
	Title           string   `gorm:"not null;size:100;uniqueIndex:idx_title_version" json:"title"`
	Description     string   `json:"description"`
	Type            SongType `gorm:"not null" json:"type"`
	OriginalSongID  *uint    `json:"-"`
	OriginalSong    *Song
	Version         int           `gorm:"default:1;uniqueIndex:idx_title_version" json:"version"`
	ReleasedAt      *time.Time    `json:"released_at"`
	Genres          []Genre       `gorm:"many2many:song_genres" json:"genres"`
	MediaSources    []MediaSource `gorm:"foreignKey:SongHashID;references:HashID;constraint:OnDelete:CASCADE,OnUpdate:CASCADE" json:"media_sources"`
	Singers         []Singer      `gorm:"many2many:song_singers" json:"singers"`
	SongSingers     []SongSinger  `gorm:"foreignKey:SongID"`
	Tags            []Tag         `gorm:"many2many:song_tags" json:"tags"`
	Aliases         []SongAlias   `gorm:"foreignKey:SongID;constraint:OnDelete:CASCADE" json:"aliases"`
	SongGroupHashID string        `gorm:"index" json:"song_group_hash_id"`
}

type SongAlias struct {
	gorm.Model
	SongID   uint   `gorm:"not null;index:idx_song_alias_name,unique" json:"-"`
	Name     string `gorm:"not null;index:idx_song_alias_name,unique" json:"name"`
	Language string `gorm:"size:8"` // ISO code

	Song Song `gorm:"foreignKey:SongID;constraint:OnDelete:CASCADE" json:"-"`
}
